libexec/functions

```diff
@@ -441,6 +441,66 @@ is_tar() {
 }
 
 
+bind_nvidia_files() {
+# either use nvidia-container-cli or nvliblist.conf to find nvidia libs and
+# binaries for binding at runtime with --nv 
+    if ! SINGULARITY_NVLIBLIST=`mktemp "${TMPDIR:-/tmp}"/.singularity-nvliblist.XXXXXXXX`; then
+        message ERROR "Failed to create temporary nvliblist file.\n"
+        ABORT 255
+    fi
+
+    # try to use nvidia-container-cli if installed and functional
+    if singularity_which nvidia-container-cli > /dev/null 2>&1 && SINGULARITY_NV_FILES=`nvidia-container-cli list -cguv`; then
+        # bins should have full paths
+        SINGULARITY_NVBINLIST=`echo "$SINGULARITY_NV_FILES" | grep -Ev "\.so"`
+        # libs should just be basenames ending in ".so"
+        for i in `echo "$SINGULARITY_NV_FILES" | grep -E "\.so"`; do
```

Comment:
is this absolutely, always entirely the case to end in so?

Reply:
It's even worse than that.  I'm just grepping to see if the lines **contain** `.so`. 

I'm doing this because I need to process libraries and binaries separately and I'm relying on the presence or absence of `.so` to determine if the file in question is a lib or bin.  

I think that the libraries are always guaranteed to have `.so` in the name.  I think the only danger is that one of the binaries ends up having `.so` in the name.  If that happens it will be included in this list, which is no big deal because this list is only used for pattern matching within the `ld.so` cache and it just won't match any pattern.  But the worse news is that it will be excluded from the binary list and won't be mounted into the container.  But that still probably isn't that big of a deal for most applications.  These bins are not being mounted into the container right now.  If a user really needs them they have to install them on their own.

The real problem is that `nvidia-container-cli` just outputs a bunch of paths with nothing to differentiate a library from a binary.  Even if they did have some kind of filtering option I wouldn't want to call `nvidia-container-cli` more than once because it is slow.

So I guess my answer here is I think there is a very small chance this could fail.  And if it fails there is an additional very small chance that failure could cause the container to fail.  But I don't know any better way to do it.

Reply:
I think that's reasonable, given what you have to work with! As long as nothing terrible, security wise could happen (other than an issue posted that we resolve) then this seems ok. Who made the `nvidia-container-cli` and why is it so doopy?

Reply:
Maybe the container client could be changed, or a command added to indicate which you want?

Reply:
> Maybe the container client could be changed, or a command added to indicate which you want?
You mean instead of defaulting to `nvidia-container-cli` the user could specify that they want to use `nvliblist.conf` instead?  I think that is a good idea.  How do you suggest I implement that?  Option + env var?

Reply:
Oh and I don't want to badmouth `nvidia-container-cli`.  It's not doopy and it's filling a big gap for us.  It is a little slow but I think that is because it is loading and unloading kernel modules?  Anyway, it is a big improvement to the guessing game that I've been playing.

Reply:
yes, **or** continue to use `nvidia-container-cli` but ask its developers to make it more customizable? Is that possible?

Reply:
Is the ask for nvidia-container-cli to have new flags like `--libraries` and `--executables` to help separate out the list of host files that should be present in the container to provide the requested capabilities?

@flx42 @3XX0

Reply:
That would be one way to do it.  But then we would have to call `nvidia-container-cli` 2x which would be slow.  Another way would be to return some indication as to the file type along with the file location in the text output.  Maybe in csv or tab delimited format or something?

Reply:
Yes this is what I had in mind! I think the two calls to be absolutely sure about the files would be worth it - the grepping will likely work for most, but I would definitely predict some error from it in the long run. Is it really significantly slow to call it once vs. twice?

Reply:
Each call to `nvidia-container-cli` takes like a second or so.  There is a noticeable pause.  So two calls will create an even more noticeable pause.  

I think it would be nice if the nvidia devs can add something like this for us to use in the future, but I don't feel it's necessary to add this to merge the PR.

Reply:
@GodloveD it shouldn't take this long, your driver is probably not setup correctly  (e.g. `nvidia-persistenced` isn't running) which forces the library to go through the full GPU initialization.

Reply:
@3XX0 Ah thanks.  That is probably the issue.  Any chance we could get some filters set up to differentiate binaries from libraries?  Should I raise an issue?

Reply:
Also @3XX0, while I have you, are there 1 or a handful of core libs that I can _always_ expect to find (regardless of driver version)?  I want to add some error checking and I thought it might be a good idea to check for the presence of one or more libs that is more or less guaranteed to be present.

Reply:
I guess for the time being we can have a filter option and we'll need to filter IPCs as well. But more generally the `list` command won't be able to follow up in terms of features so we will need to fork and implement the library directly in Singularity.

Regarding your second comment, unfortunately there is no way of knowing.
This why the library has relaxed checks right now and several options to tweak the driver features. I was planning on implementing a strict mode which checks things more thoroughly though. Also, this list will likely change in future driver versions.

Reply:
@vsoch 
>Maybe the container client could be changed, or a command added to indicate which you want?

I added an env var (`SINGULARITY_USE_NVLIBLIST_CONF`) so that users can ignore `nvidia-container-cli` and use nvliblist.conf if that better suits their needs.